Source/JavaScriptCore/b3/air/AirAllocateRegistersAndStackAndGenerateCode.cpp

```diff
@@ -211,14 +239,22 @@ ALWAYS_INLINE void GenerateAndAllocateRegisters::alloc(Tmp tmp, Reg reg, Arg::Ro
     if (Arg::isAnyUse(role)) {
         JIT_COMMENT(*m_jit, "Alloc(", tmp, ", ", reg, ", role=", role, ")");
         intptr_t offset = m_map[tmp].spillSlot->offsetFromFP();
-        if (tmp.bank() == GP)
-            m_jit->loadRegWord(callFrameAddr(Air::Move, *m_jit, offset), reg.gpr());
-        else if (B3::conservativeRegisterBytes(B3::FP) == sizeof(double) || !m_code.usesSIMD()) {
+        if (tmp.bank() == GP) {
+            auto src = callFrameAddr(Air::Move, *m_jit, offset, m_code.frameSize(), registerWidth());
+            if (src.isAddr())
+                m_jit->loadRegWord(src.asAddress(), reg.gpr());
+            else
+                m_jit->loadRegWord(src.asBaseIndex(), reg.gpr());
+        } else if (B3::conservativeRegisterBytes(B3::FP) == sizeof(double) || !m_code.usesSIMD()) {
             ASSERT(m_map[tmp].spillSlot->byteSize() == bytesForWidth(Width64));
-            m_jit->loadDouble(callFrameAddr(Air::MoveDouble, *m_jit, offset), reg.fpr());
+            auto src = callFrameAddr(Air::MoveDouble, *m_jit, offset, m_code.frameSize(), Width64);
+            ASSERT(src.isAddr());
```

Comment:
Can you explain why this is correct?

Reply:
It used to be an invariant because this patch `isValidForm` rejected BaseIndex addresses for Air::MoveDouble. I've changed `isValidIndexForm` in https://github.com/WebKit/WebKit/commit/da11283e6bd1aaa4c73f84c839f1a7b72f508ef3 to only reject  Air::MoveDouble and Air::MoveFloat on ARM_THUMB2 (because the MacroAssemblerARMv7 will try to use the addressTempRegister to materialize them.) Hence, I've replaced the ASSERTs with branches to handle BaseIndex addresses for loadDouble/storeDouble/loadVector/storeVector (those can and do occur on ARM64.)